Sync `link` specific UA stylesheet rule as per SVG2 Web Specification

Ahmad-S792 opened the pull request:
#### 042030db2bb825fe23c6d49310fa6e37e82243f8
<pre>
Sync `link` specific UA stylesheet rule as per SVG2 Web Specification

<a href="https://bugs.webkit.org/show_bug.cgi?id=269076">https://bugs.webkit.org/show_bug.cgi?id=269076</a>

Reviewed by Anne van Kesteren.

This patch aligns WebKit with Gecko / Firefox, Blink / Chromium and Web Specification:

[1] <a href="https://svgwg.org/svg2-draft/styling.html#UAStyleSheet">https://svgwg.org/svg2-draft/styling.html#UAStyleSheet</a>

As per SVG2 specification, this PR adds UA rules for links wrapped in SVG
element to have cursor as `pointer`.

Additionally, this PR also sync test to test above behavior from WPT:

Upstream commit: <a href="https://github.com/web-platform-tests/wpt/commit/43a23d0263d9d8a3c0fad2c28d8ff608f4343bb9">https://github.com/web-platform-tests/wpt/commit/43a23d0263d9d8a3c0fad2c28d8ff608f4343bb9</a>

* Source/WebCore/css/svg.css:
(a:any-link):
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/the-css-user-agent-style-sheet-and-presentational-hints/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/the-css-user-agent-style-sheet-and-presentational-hints/no-help-cursor-on-links-wrapped-in-svg.historical.html:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/the-css-user-agent-style-sheet-and-presentational-hints/no-help-cursor-on-links-wrapped-in-svg.historical-expected.txt:

Canonical link: <a href="https://commits.webkit.org/274422@main">https://commits.webkit.org/274422@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b394d312aa239c49fc31cd8391409a3332e55702

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39062 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17995 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41415 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41596 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34779 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41368 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20884 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15343 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32688 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15164 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33857 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13166 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13126 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34792 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42873 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35460 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35120 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38954 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13874 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11439 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37182 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15480 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/34071 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8739 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15142 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14966 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/css/svg.css:
```diff
@@ -25,7 +25,10 @@
+/* https://svgwg.org/svg2-draft/styling.html#UAStyleSheet */
+
 @namespace "http://www.w3.org/2000/svg";
+@namespace xml "http://www.w3.org/XML/1998/namespace";
```
I don't think we should add this until it's actually needed for something.

annevk commented on Source/WebCore/css/svg.css:
```diff
@@ -84,3 +87,7 @@ text, tspan, tref {
+
+:link, :visited {
+    cursor: pointer;
+}
```
In `html.css` we have this (which looks weird, to be fair):

```
a:any-link {
    color: -webkit-link;
    text-decoration: underline;
    cursor: auto;
}

a:any-link:active {
    color: -webkit-activelink;
}

area:any-link {
    cursor: pointer;
}
```
I would expect these to be aligned in some way.

I think we want to at least use `:any-link` and tell the SVG WG to adopt that as well.

Ahmad-S792 replied:
Yes - I also checked both UA stylesheet for Firefox and Chrome and they have:

```
a:any-link {
    cursor: pointer; <- Only chrome has this; Firefox has `text-decoration: none`.
}

```
Should I change to above?

NOTE - our `html.css` should have also `cursor: pointer`. We fail WPT due to it, also @vitorroriz tried it in past but it started causing cursor issue on Amazon Prime.

Ahmad-S792 replied:
Plus if I look into HTML Spec as well, we have:

```:link, :visited { text-decoration: underline; cursor: pointer; }```

https://html.spec.whatwg.org/multipage/rendering.html#phrasing-content-3

Ahmad-S792 replied:
@annevk - https://github.com/whatwg/html/issues/10133

annevk replied:
Thanks, if we make this `:any-link` and remove the additional namespace declaration I think I'm okay with trying this out. I'd be interested in learning about the Amazon issue. Maybe something we need to raise with whatwg/html as well.

Ahmad-S792 replied:
> Thanks, if we make this `:any-link` and remove the additional namespace declaration I think I'm okay with trying this out. I'd be interested in learning about the Amazon issue. Maybe something we need to raise with whatwg/html as well.

@annevk - I made the changes and the @vitorroriz bug is here - https://bugs.webkit.org/show_bug.cgi?id=173909 and this comment from Brent (might be Radar link as well below it) would give more context - https://bugs.webkit.org/show_bug.cgi?id=173909#c6

Ahmad-S792 committed:
Sync `link` specific UA stylesheet rule as per SVG2 Web Specification

https://bugs.webkit.org/show_bug.cgi?id=269076

Reviewed by Anne van Kesteren.

This patch aligns WebKit with Gecko / Firefox, Blink / Chromium and Web Specification:

[1] https://svgwg.org/svg2-draft/styling.html#UAStyleSheet

As per SVG2 specification, this PR adds UA rules for links wrapped in SVG
element to have cursor as `pointer`.

Additionally, this PR also sync test to test above behavior from WPT:

Upstream commit: https://github.com/web-platform-tests/wpt/commit/43a23d0263d9d8a3c0fad2c28d8ff608f4343bb9

* Source/WebCore/css/svg.css:
(a:any-link):
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/the-css-user-agent-style-sheet-and-presentational-hints/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/the-css-user-agent-style-sheet-and-presentational-hints/no-help-cursor-on-links-wrapped-in-svg.historical.html:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/the-css-user-agent-style-sheet-and-presentational-hints/no-help-cursor-on-links-wrapped-in-svg.historical-expected.txt:

Canonical link: https://commits.webkit.org/274422@main
